sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -699,8 +699,80 @@ case class UnionExec(children: Seq[SparkPlan]) extends SparkPlan {
     }
   }
 
-  protected override def doExecute(): RDD[InternalRow] =
-    sparkContext.union(children.map(_.execute()))
+  /**
+   * Returns the output partitionings of the children, with the attributes converted to
+   * the first child's attributes at the same position.
+   */
+  private def prepareOutputPartitioning(): Seq[Partitioning] = {
+    val attributesMap = children.map(_.output).transpose.flatMap { attrs =>
```

Comment:
Maybe separate maps per columns would be safer if `output` of some of the `children.tail` contains duplicates and the those duplicates map to different attributes from `children.head.output`?